Migrate low level core API changes to core.

msiglreith opened the pull request:
Add Surface, Instance, Swapchain, CommandQueue, QueueFamily and Adapter traits to core from corell.

The backends would need updates later-on. The dx12ll and vulkanll backends can be used for reference.
After the backend changes, the window crates shouldn't be needed anymore and replaced by the `Surface` and `SwapChain` parts.

Open question would be the support for other windows like SDL and glfw, because the current surface trait uses an associated window type.

Bastacyclop commented:
Could the window crates handle the `Instance` creation, each having their own types ?

msiglreith commented:
Oh, indeed this probably works fine as well.

kvark commented on src/core/src/lib.rs:
```diff
@@ -238,3 +239,107 @@ pub trait Device: Sized {
+/// An `Instance` holds per-application state for a specific backend
+pub trait Instance {
+    /// Associated `Adapter` type.
+    type Adapter: Adapter;
```
is there a single adapter associated with the instance?
I thought the instance is the main entry point, and it can have multiple different adapters exposed.
These adapters may have different queue types/families, so must have different types.

msiglreith replied:
There is a single adapter type associated, which is more like a backend abstraction for all adapters.
Adapters then define hardware devices and virtual devices and there capabilities (supported queue types).
I don't really see the issue here unfortunately.

kvark replied:
Why should we have `Instance` and `Adapter` be separate types? They both kinda converge to representing a backend.

kvark commented on src/core/src/lib.rs:
```diff
@@ -238,3 +239,107 @@ pub trait Device: Sized {
+    fn get_info(&self) -> &AdapterInfo;
+
+    /// Return the supported queue families for this adapter.
+    fn get_queue_families(&self) -> Iter<Self::QueueFamily>;
```
any reason it can't just return `&[Self::QueueFamily]`?
Also, why is `QueueFamily` a trait?

msiglreith replied:
I considered it easier to use to return `Iter` instead of the slice. (It's basically the iterator equivalent to &[]).
Will change it if you prefer returning a slice.

`QueueFamily` is defined as trait as it was easier to implement and wasn't sure if it can be implemented as single struct for all backends. I can try to define a struct for it. Difficulties arised because of 
checking for presentation support for a surface.

kvark replied:
If it's always an array/slice on the implementation side, I'd strongly be for `&[]` returned value since it has less magic than returning an iterator.
As for the `QueueFamily` trait - I suppose it's fine, especially since most apps will not care about it.

kvark commented on src/core/src/lib.rs:
```diff
@@ -238,3 +239,107 @@ pub trait Device: Sized {
+    type SwapChain: SwapChain;
+
+    /// Create a new swapchain from the current surface with an associated present queue.
+    fn build_swapchain<T: format::RenderFormat>(&self, present_queue: &Self::CommandQueue)
```
I wonder... if a surface is only used to build the swap chain, why do we need the concept of a surface at all? Can't the instance return a swap chain directly, bypassing the surface (or hiding it)?

msiglreith replied:
I would also prefer to remove it. Similar issue as above: require a queue, which supports presentation on a surface to create a swapchain. But I need the surface first to check presentation support.
Again, I'll re-think it to see if I come up with a better solution!

kvark commented:
Thanks @msiglreith ! I think this has a few rough edges but this is totally fine for a WIP patch. Ready to merge when you are.

msiglreith commented:
Replaced `Iter` with `&[]` and split the `Instance` trait into separate functions for `Adapter` and `Surface`.
This might make a little bit more difficult to implement for the backends but hopefully improves usability.

kvark commented:
Sounds good, thanks!
@homu r+

homu commented:
:pushpin: Commit 5ba6651 has been approved by `kvark`

<!-- @homu r=kvark 5ba665113f40e60f1ea83761d6e0e47e447d655b -->

homu commented:
:hourglass: Testing commit 5ba6651 with merge 26d2a55...

homu commented:
:boom: Test timed out

kvark commented:
@homu retry

homu commented:
:hourglass: Testing commit 5ba6651 with merge 798679f...

homu commented:
:boom: Test timed out

kvark commented:
Travis CI is having a hard time with OSX bots at the moment... Could anyone confirm that this PR builds on OSX, so that we can just force-merge it?

kvark commented:
@homu retry

> On Jan 31, 2017, at 23:18, Homu <notifications@github.com> wrote:
> 
> 💥 Test timed out
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

homu commented:
:hourglass: Testing commit 5ba6651 with merge c87d35b...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/gfx-rs/gfx/builds/197494701)

JohnColanduoni commented:
Builds fine on OS X 10.12
